File: Swift/거리두기_확인하기.swift
import Foundation

func solution(_ places:[[String]]) -> [Int] {
    var res = [Int]()
    for place in places {
        var target = [(Int, Int)]()
        for (i, row) in place.enumerated() {
            for (k, col) in row.enumerated() {
                if col == "P" {
                    target.append((i, k))
                }
            }
        }
        
        var isFailed = false
        for i in target.indices {
            if isFailed { break }
            for k in (i+1)..<target.count {
                let cur = target[i]
                let next = target[k]
                let dist = abs(cur.0 - next.0) + abs(cur.1 - next.1)
                
                if dist > 2 { continue }
                else if dist == 1 {
                    isFailed = true
                    break
                }
                else if dist == 2 {
                    if cur.0 == next.0 {
                        if place[cur.0][min(cur.1, next.1) + 1] == "O" {
                            isFailed = true
                            break
                        }
                    }
                    else if cur.1 == next.1 {
                        if place[min(cur.0, next.0) + 1][cur.1] == "O" {
                            isFailed = true
                            break
                        }
                    }
                    else {
                        if place[cur.0][next.1] == "O" || place[next.0][cur.1] == "O" {
                            isFailed = true
                            break
                        }
                    }
                }
                
            }
        }
        res.append(isFailed ? 0 : 1)
    }
    return res
}

extension String {
    subscript(_ index: Int) -> Character {
        self[self.index(startIndex, offsetBy: index)]
    }
}

File: Python/거리두기_확인하기.py
def solution(places):
    answer = [1, 1, 1, 1, 1]
    for idx, place in enumerate(places):
        p_index = []
        near = []
        for i, p in enumerate(place):
            p_index += [(i, ii) for ii, s in enumerate(p) if s == 'P']
        for i, l in enumerate(p_index):
            for r in p_index[i + 1:]:
                if abs(l[0] - r[0]) + abs(l[1] - r[1]) <= 2:
                    near += [[l, r]]

        for n in near:
            if n[0][0] == n[1][0]:
                if n[0][1] - n[1][1] == -1:
                    answer[idx] = 0
                    break
                elif n[0][1] - n[1][1] == -2 and place[n[0][0]][n[0][1] + 1] != 'X':
                    answer[idx] = 0
                    break

            if n[0][1] == n[1][1]:
                if n[0][0] - n[1][0] == -1:
                    answer[idx] = 0
                    break
                elif n[0][0] - n[1][0] == -2 and place[n[0][0] + 1][n[0][1]] != 'X':
                    answer[idx] = 0
                    break
            if n[0][1] != n[1][1] and n[0][0] != n[1][0]:
                if n[0][1] > n[1][1] and (place[n[0][0] + 1][n[0][1]] != 'X' or place[n[0][0]][n[0][1] - 1] != 'X'):
                    answer[idx] = 0
                    break
                if n[0][1] < n[1][1] and (place[n[0][0] + 1][n[0][1]] != 'X' or place[n[0][0]][n[0][1] + 1] != 'X'):
                    answer[idx] = 0
                    break
    return answer